lib/src/test/java/io/ably/lib/util/CryptoMessageTest.java

```diff
@@ -0,0 +1,153 @@
+package io.ably.lib.util;
+
+import static io.ably.lib.test.common.Helpers.assertMessagesEqual;
+import static org.hamcrest.core.StringEndsWith.endsWith;
+import static org.junit.Assert.assertArrayEquals;
+import static org.junit.Assert.assertEquals;
+import static org.junit.Assert.assertThat;
+
+import java.io.IOException;
+import java.security.NoSuchAlgorithmException;
+
+import org.junit.Test;
+import org.junit.runner.RunWith;
+import org.junit.runners.Parameterized;
+import org.junit.runners.Parameterized.Parameters;
+
+import io.ably.lib.test.common.Setup;
+import io.ably.lib.types.AblyException;
+import io.ably.lib.types.ChannelOptions;
+import io.ably.lib.types.Message;
+import io.ably.lib.util.Base64Coder;
+import io.ably.lib.util.Crypto;
+import io.ably.lib.util.Crypto.CipherParams;
+
+@RunWith(Parameterized.class)
+public class CryptoMessageTest {
+	public enum FixtureSet {
+		AES128(16),
+		AES256(32);
+
+		public final byte[] key;
+		public final byte[] iv;
+		private final String fileName;
+		public final String cipherName;
+
+		private FixtureSet(final int keySize) {
+			if (keySize < 1) {
+				throw new IllegalArgumentException("keySize");
+			}
+
+			final int keyLength = keySize * 8; // bytes to bits
+			fileName = "crypto-data-" + keyLength;
+			cipherName = "cipher+aes-" + keyLength + "-cbc";
+
+			CryptoTestData testData;
+			try {
+				testData = loadTestData();
+			} catch (IOException e) {
+				throw new Error(e); // caught to uncaught
+			}
+			key = Base64Coder.decode(testData.key);
+			iv = Base64Coder.decode(testData.iv);
+
+			if (keySize != this.key.length) {
+				throw new IllegalArgumentException("key");
+			}
+			if (16 != this.iv.length) {
+				throw new IllegalArgumentException("iv");
+			}
+		}
+
+		private CryptoTestData loadTestData() throws IOException {
+			return (CryptoTestData)Setup.loadJson(
+				"ably-common/test-resources/" + fileName + ".json",
+				CryptoTestData.class);
+		}
+	}
+
+	@Parameters(name= "{0}")
+	public static Object[][] data() {
+		return new Object[][] {
+			{ FixtureSet.AES128 },
+			{ FixtureSet.AES256 },
+		};
+	}
+
+	private final FixtureSet fixtureSet;
+
+	public CryptoMessageTest(final FixtureSet fixtureSet) {
+		this.fixtureSet = fixtureSet;
+	}
+
+	@Test
+	public void testDecrypt() throws NoSuchAlgorithmException, CloneNotSupportedException, AblyException, IOException {
+		final CryptoTestData testData = fixtureSet.loadTestData();
+		final String algorithm = testData.algorithm;
+
+		final CipherParams params = Crypto.getParams(algorithm, fixtureSet.key, fixtureSet.iv);
+		final ChannelOptions options = new ChannelOptions() {{encrypted = true; cipherParams = params;}};
+
+		for(final CryptoTestItem item : testData.items) {
+			final Message plain = item.encoded;
+			final Message encrypted = item.encrypted;
+			assertThat(encrypted.encoding, endsWith(fixtureSet.cipherName + "/base64"));
+
+			// if necessary, remove base64 encoding from plain-'text' message
```

Comment:
Comment syntax is inconsistent with other changes in this commit. In this repo we've been using `/* .... */`-style comments generally.

Reply:
Thanks. I'm going to leave this as is - it'll get picked up, if necessary, when we get around to conforming code style across the codebase. For now I am happy that, albeit in a minority, my commenting style isn't entirely non-standard even within this codebase:

```
ably-java % git grep '/\* ' | wc -l
    2332
ably-java % git grep '// ' | wc -l
     191
```

Reply:
Issue created: #537